### PR TITLE
raftstore-v2: memtable pre-flush should be async

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #14566

What's Changed:

```commit-message
memtable pre-flush should be async
```
Sync flush may lead to waiting in lines if there are many splits concurrently.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tabokie — COMMENTED** [review]
What's wrong with waiting in line? If split request is only sent after the flush returns?

**Comment by SpadeA-Tang** [comment]
> What's wrong with waiting in line? If split request is only sent after the flush returns?

This can only be guaranteed for leader. Pre flush for follower is just a notification.

**Comment by tabokie** [comment]
Suppose A and B are located on the same store, A's follower flush will still block B's leader flush even if it's async, because they both share the same thread pool.

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 1]
```diff
@@ -63,12 +63,13 @@ impl<EK: KvEngine, ER: RaftEngine> Runner<EK, ER> {
             .get(region_id)
             .map(|mut cache| cache.latest().cloned()) else {return};
         let now = Instant::now();
-        tablet.flush_cfs(DATA_CFS, true).unwrap();
+        // sync flush for leader to let the flush happend before later checkpoint.
+        tablet.flush_cfs(DATA_CFS, is_leader).unwrap();
         let elapsed = now.saturating_elapsed();
         // to be removed after when it's stable
         info!(
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet_flush.rs:70]: how about use metrics?

**Comment by SpadeA-Tang** [comment]
> Suppose A and B are located on the same store, A's follower flush will still block B's leader flush even if it's async, because they both share the same thread pool.

Blocking leader's flush is fine as the split is still not proposed. There's possibility that leader's flush may block later follower's flush, maybe future worker is more appropriate here?

File: components/raftstore-v2/src/worker/tablet_flush.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet_flush.rs:70]: It's temporarily used here to debug and diagnose.

**Comment by tabokie** [comment]
> There's possibility that leader's flush may block later follower's flush

This PR doesn't help this case either. Follower's flush will be immediately scheduled but not executed. Its latency will be the same async or not.

> maybe future worker is more appropriate here?

It doesn't matter what worker you use in TiKV. All the actual work happens in RocksDB's own thread pool.

**Comment by SpadeA-Tang** [comment]
> This PR doesn't help this case either. Follower's flush will be immediately scheduled but not executed. Its latency will be the same async or not.

If rocksdb have, say 4 (or more) threads for flush, but tikv only has 2 workers for this, sync may not utilize the other two threads.

**Comment by SpadeA-Tang** [comment]
> Additionally, I don't know how this is related to the phenomenon we saw earlier (concurrent flush causes QPS drop), because based on this code, split is only proposed after flush is completed.

The problem is not occurred for leader, it's for follower. Leader's pre flush strictly happens before split propose, so the checkpoint is lightweight and can complete quickly. However, for follower, due to sync flush, there's may be a waiting in line for flush. It means the flush may not start even when it is to apply the split. So the checkpoint is still heavy work for followers. I think that make follower async flush and turn rocksdb's flush thread numbers may help.

**Review by tabokie — COMMENTED** [review]
> If rocksdb have, say 4 (or more) threads for flush, but tikv only has 2 workers for this, sync may not utilize the other two threads.

This is true. But your way of implementation is confusing, because you are still doing sync flush in this worker. Follower flush will still be blocked if leader flush arrives first. A better way is to move sync flush to another poo, just like tablet_gc::Runner::background_pool.

**Comment by SpadeA-Tang** [comment]
> This is true. But your way of implementation is confusing, because you are still doing sync flush in this worker. Follower flush will still be blocked if leader flush arrives first. A better way is to move sync flush to another pool, just like tablet_gc::Runner::background_pool.

Sync leader flush blocking async follower flush is indeeded what I concerneded. And thansk for the good suggestion.

**Review by tabokie — COMMENTED** [review]
You can unify the two workers:
Add an optional callback field to tablet_gc::TabletFlush. If the callback is none, then tablet_gc will call flush immediately with sync=false, if the callback is present, tablet_gc will schedule the flush to run in the background pool.

**Comment by SpadeA-Tang** [comment]
> You can unify the two workers: Add an optional callback field to tablet_gc::TabletFlush. If the callback is none, then tablet_gc will call flush immediately with sync=false, if the callback is present, tablet_gc will schedule the flush to run in the background pool.

Good suggestion.

File: components/raftstore-v2/src/worker/tablet.rs [hunk 2]
```diff
@@ -41,6 +42,13 @@ pub enum Task<EK> {
     DirectDestroy { tablet: Either<EK, PathBuf> },
     /// Cleanup ssts.
     CleanupImportSst(Box<[SstMeta]>),
+    /// Flush memtable before split
+    ///
+    /// split_req is some iff the task is sent from leader
+    TabletPreFlush {
```
Comment (tabokie) [inline components/raftstore-v2/src/worker/tablet.rs:48]:
```suggestion
    Flush {
```

**Review by tabokie — APPROVED** [review]

File: components/raftstore-v2/src/worker/tablet.rs [hunk 3]
```diff
@@ -41,6 +42,13 @@ pub enum Task<EK> {
     DirectDestroy { tablet: Either<EK, PathBuf> },
     /// Cleanup ssts.
     CleanupImportSst(Box<[SstMeta]>),
+    /// Flush memtable before split
+    ///
+    /// on_flush_finish is some iff the task is sent from leader
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet.rs:47]:
```suggestion
    /// on_flush_finish can't be none if the task is sent from leader
```

File: components/raftstore-v2/src/worker/tablet.rs [hunk 4]
```diff
@@ -41,6 +42,13 @@ pub enum Task<EK> {
...
+    Flush {
+        region_id: u64,
+        on_flush_finish: Option<Box<dyn FnOnce() + Send>>,
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet.rs:50]: how about rename "callback".

File: components/raftstore-v2/src/worker/tablet.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet.rs:47]: "iff" should be used here as it cannot be some when sent from follower.

File: components/raftstore-v2/src/worker/tablet.rs [hunk 5]
```diff
@@ -77,6 +85,17 @@ impl<EK> Display for Task<EK> {
             Task::CleanupImportSst(ssts) => {
                 write!(f, "cleanup import ssts {:?}", ssts)
             }
+            Task::Flush {
+                region_id,
+                on_flush_finish,
+            } => {
+                write!(
+                    f,
+                    "pre-flush tablet for region_id {}, is leader {}",
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet.rs:94]: maybe needs to change?

File: components/raftstore-v2/src/worker/tablet.rs [hunk 6]
```diff
@@ -311,6 +330,46 @@ impl<EK: KvEngine> Runner<EK> {
             }
         }
     }
+
+    fn pre_flush_tablet(&self, region_id: u64, on_flush_finish: Option<Box<dyn FnOnce() + Send>>) {
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet.rs:334]: how about rename `flush` direct?

File: components/raftstore-v2/src/worker/tablet.rs (hunk 3 above)
Comment (bufferflies) [reply components/raftstore-v2/src/worker/tablet.rs:47]: And you should add some check in follower？

File: components/raftstore-v2/src/worker/tablet.rs [hunk 7]
```diff
@@ -311,6 +331,46 @@ impl<EK: KvEngine> Runner<EK> {
...
+            self.background_pool
+                .spawn(async move {
+                    let now = Instant::now();
+                    // sync flush for leader to let the flush happend before later checkpoint.
+                    tablet.flush_cfs(DATA_CFS, true).unwrap();
+                    let elapsed = now.saturating_elapsed();
+                    // to be removed after when it's stable
+                    info!(
+                        logger,
+                        "pre-flush memtable for leader";
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet.rs:355]:
```suggestion
                        "flush memtable for leader";
```

File: components/raftstore-v2/src/worker/tablet.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet.rs]: Runner have no other information to see if it's follower or leader. Some or None is enough.

**Pull request merged**